### ISSUE TITLE
Unit calculation error

### ISSUE THREAD

**Issue by idealkindom** [Proektsoftbg/Calcpad#137]
I paste the calculation txt here:
ρ = 1.6 kcal/cm^3

J = 1000 kcal/kg

ρ/J = 1.6 kcal/cm^3/1000 kcal/kg = 3.82×10-7 (4186800·t)/m3

**Comment by Proektsoftbg** [comment]
Hi! Thank you for reporting this. Probably the result is internally correct, but Calcpad does not know how to compose and display the target units. If you add the target units at the end, say `|kg/cm^3`, it will convert accordingly:
```
ρ = 1.6kcal/cm^3
J = 1000kcal/kg
ρ/J|kg/cm^3 '= 0.0016 kg/cm3
```
![image](https://user-images.githubusercontent.com/90918363/225033669-11d1b5c0-d98c-449a-9052-206f054e5986.png)

We have taught Calcpad to properly compose target units for mechanical and electrical engineering. For other fields, we will need assistance by others. However, you can always force the desired target units by adding vertical bar "`|`".

**Comment by Proektsoftbg** [comment]
Yes, it really looks strange, but this is because kcal is not a SI unit and it internally incudes this "strange" number:

`1kcal|J` will give: 1kcal = 4186.8 J

Calcpad, like other numerical packages, decomposes the units to vectors of factors and powers. Each element in the vector corresponds to one of the basic 7 physical dimensions. Then, it performs all the arithmetic operations on the vectors and finally composes back the resulting units. You can learn more about how units work in Calcpad from this blog post:

https://wordpress.com/post/calcpad.blog/99

If you use the SI units `J`, instead of `kcal`, you will see that the "strange" number disappears like magic, and the result looks right straight away:

![image](https://user-images.githubusercontent.com/90918363/225053778-2d194aa6-9308-4778-b825-1d2d93b0df83.png)

Theoretically, it is possible to avoid this problem by using some kind of symbolic evaluation. However, Calcpad still do not have a symbolic engine. Also, symbolic calculations are much slower than numerical ones. So, specifying the target units explicitly in some cases may be needed and it is not a trouble.

And probably, you will also ask: Why does it keep this strange number in the units and does not let it go in the result.
The answer is to allow calculations in USCS or UK units without falling down to SI units on each formula, e.g.:

![image](https://user-images.githubusercontent.com/90918363/225064278-23c9c39c-23e4-4e55-9a92-6b5675e6eef5.png)

These units also include such internal factors. However, I will think about some workaround. Probably, I will be able to figure out something. So, thank you once again for your feedback. It is useful.

**Comment by idealkindom** [comment]
> Yes, it really looks strange, but this is because kcal is not a SI unit and it internally incudes this "strange" number:
> 
> `1kcal|J` will give: 1kcal = 4186.8 J
> 
> Calcpad, like other numerical packages, decomposes the units to vectors of factors and powers. Each element in the vector corresponds to one of the basic 7 physical dimensions. Then, it performs all the arithmetic operations on the vectors and finally composes back the resulting units. You can learn more about how units work in Calcpad from this blog post:
> 
> https://calcpad.blog/2018/01/28/how-units-work/
> 
> If you use the SI units `J`, instead of `kcal`, you will see that the "strange" number disappears like magic, and the result looks right straight away:
> 
> ![image](https://user-images.githubusercontent.com/90918363/225053778-2d194aa6-9308-4778-b825-1d2d93b0df83.png)
> 
> Theoretically, it is possible to avoid this problem by using some kind of symbolic evaluation. However, Calcpad still do not have a symbolic engine. Also, symbolic calculations are much slower than numerical ones. So, specifying the target units explicitly in some cases may be needed and it is not a trouble.
> 
> And probably, you will also ask: Why does it keep this strange number in the units and does not let it go in the result. The answer is to allow calculations in USCS or UK units without falling down to SI units on each formula, e.g.:
> 
> ![image](https://user-images.githubusercontent.com/90918363/225064278-23c9c39c-23e4-4e55-9a92-6b5675e6eef5.png)
> 
> These units also include such internal factors. However, I will think about some workaround. Probably, I will be able to figure out something. Until then, please use the manual conversion. Thank you once again for your feedback. It was useful.

Thanks for your concrete explanation.Now I understand the.reason behind the output. Maybe this could be improved by intrducing some algotithm, which preprocess all the unist.ahead before.it do the calculation, thus allowing to erase same unit at first.

**Comment by Proektsoftbg** [comment]
Yes, but it will be more likely a kind of post-processing. It will be easier and faster. We will have to check if the residual factors are convertible to any letter-only prefixes like "kilo", "mega" or non-SI units like "lbf" or "kip" for mass or "ft", "in" for length, etc. If the factors do not match any of them, to reduce the respective factor to the nearest convertible and multiply the answer by the rest.

In your case, it will multiply 3.82*10^-7*4186800 = 1.6 in t/m^3.

But I will do that a bit further because I have a lot of other work these days. Need to finish some urgent projects now.

**Comment by idealkindom** [comment]
> Yes, but it will be more likely a kind of post-processing. It will be easier and faster. We will have to check if the residual factors are convertible to any letter-only prefixes like "kilo", "mega" or non-SI units like "lbf" or "kip" for mass or "ft", "in" for length, etc. If the factors do not match any of them, to reduce the respective factor to the nearest convertible and multiply the answer by the rest.
> 
> In your case, it will multiply 3.82_10^-7_4186800 = 1.6 in t/m^3.
> 
> But I will do that a bit further because I have a lot of other work these days. Need to finish some urgent projects now.

Thanks for your great job,Ned.

**Comment by Proektsoftbg** [comment]
Hi!

I fixed it in version 5.9.9:
https://github.com/Proektsoftbg/Calcpad/releases/tag/v5.9.9

After calculations, units are normalized as discussed above. Any residual factors are transferred to the result and not vidiblr in the units anymore.
